### PR TITLE
feat(desktop): finalize slash autocomplete acceptance behavior

### LINKED ISSUE #343: [S03] Acceptance + insertion behavior
Tab/Enter inserts command + space; Esc dismisses.

# S03: Acceptance + insertion behavior

**Goal:** Complete slash autocomplete acceptance so selecting a suggestion commits command text into the composer with trailing space, closes suggestions, and keeps typing flow uninterrupted.

**Demo:** Type `/ka`, highlight `/kata plan`, press `Tab` (or `Enter`) — textarea becomes `/kata plan `, dropdown closes, focus remains in the textarea, and user can immediately type arguments.

## Requirement Coverage (owned/supported)

- **Owned:** R006 — accepting a suggestion inserts command + trailing space.
- **Supported:** R005 (Tab/Enter acceptance while navigating), R007 (Esc dismisses without insertion), R001 (autocomplete interaction remains slash-triggered and stable during acceptance).

## Must-Haves

- `Tab` and `Enter` both accept the currently highlighted suggestion (R005 + R006)
- Accepted suggestion inserts `command + " "` exactly once (no double insertion)
- Dropdown closes immediately after accept or Esc dismiss
- Textarea focus and caret remain in-input at end of inserted command
- Existing send behavior (`Enter` submits) still works when dropdown is closed

## Proof Level

- This slice proves: **integration** (keyboard acceptance contract wired through dropdown + MessageInput runtime path)
- Real runtime required: **yes** (DOM focus/caret and keyboard arbitration behavior)
- Human/UAT required: **no** (automated assertions + targeted runtime verification are sufficient)

## Verification

- `bun run test -- src/renderer/components/chat/__tests__/MessageInput.slash-acceptance.test.tsx`
- `bun run test -- src/renderer/components/chat/__tests__/CommandSuggestionDropdown.acceptance.test.tsx`
- `bun run typecheck`
- Runtime check: `bun run build && bun run start` then verify `/` → arrow selection → `Tab` inserts with trailing space; repeat with `Enter`; `Esc` dismisses without insertion.
- Diagnostic check: tests assert `SLASH_ACCEPT_NO_SELECTION` warning path when acceptance key is pressed with no active suggestion.

## Observability / Diagnostics

- Runtime signals: structured dev-only diagnostics under `[SlashAutocomplete]` with stable codes (`SLASH_ACCEPTED`, `SLASH_ACCEPT_NO_SELECTION`, `SLASH_ACCEPT_SUPPRESSED_DUPLICATE`).
- Inspection surfaces: `MessageInput` acceptance state is testable via DOM value/caret assertions and emitted diagnostic code paths.
- Failure visibility: failed acceptance arbitration emits stable code + key + current input prefix (no sensitive payloads).
- Redaction constraints: command tokens are non-sensitive; never log full freeform message text outside slash prefix context.

## Integration Closure

- Upstream surfaces consumed: S02-produced `CommandSuggestionDropdown` and `useCommandSuggestions` contracts, plus existing `MessageInput` submit handler.
- New wiring introduced in this slice: deterministic accept-on-key path in `MessageInput`, callback handshake between highlighted suggestion and textarea value/caret management, and acceptance diagnostics.
- What remains before the milestone is truly usable end-to-end: S04 regression coverage expansion, docs updates, and UAT report capture.

## Dependency Notes (S02)

- Dependency metadata readback: S02 resolves to issue `#342` with `dependsOn:[S01]` and is structurally resolvable.
- Forward-intelligence carryover: acceptance must not regress S02 arrow/Esc behavior; Enter key arbitration is the fragile edge because MessageInput currently uses Enter-to-send.

## Tasks

- [ ] **T01: Create failing acceptance contract tests for command insertion** `est:40m`
  - Why: Lock expected behavior before implementation so S03 can prove R006 and guard against double-insertion regressions.
  - Files: `src/renderer/components/chat/__tests__/MessageInput.slash-acceptance.test.tsx`, `src/renderer/components/chat/__tests__/CommandSuggestionDropdown.acceptance.test.tsx`
  - Do:
    1. Add MessageInput integration tests covering Tab/Enter acceptance, trailing-space insertion, focus retention, and Esc dismiss behavior.
    2. Add dropdown acceptance tests covering onSelect key paths and no-op behavior when no selection exists.
    3. Include an explicit failing assertion for duplicate insertion prevention.
  - Verify: `bun run test -- src/renderer/components/chat/__tests__/MessageInput.slash-acceptance.test.tsx` (expected red initially).
  - Done when: tests exist with meaningful assertions and fail for missing S03 behavior.

- [ ] **T02: Implement deterministic accept-suggestion insertion in MessageInput** `est:55m`
  - Why: S03’s core user-visible behavior is insertion + trailing space + caret/focus continuity under Tab/Enter acceptance.
  - Files: `src/renderer/components/chat/MessageInput.tsx`, `src/renderer/hooks/useCommandSuggestions.ts`
  - Do:
    1. Add a single acceptance path that consumes the currently selected suggestion and writes `name + " "` to textarea state.
    2. Ensure caret lands at the end and textarea focus is retained after insertion.
    3. Suppress Enter-to-send when dropdown is open and acceptance is handled.
    4. Prevent duplicate insertion across keydown/select callback interplay.
  - Verify: targeted acceptance tests move from red to green for insertion/caret cases.
  - Done when: Tab/Enter acceptance updates input exactly once with trailing space and preserves focus/caret.

- [ ] **T03: Finalize keyboard arbitration + stable diagnostics and close S03 verification** `est:40m`
  - Why: Complete reliable key routing and observable failure paths so future agents can localize acceptance bugs quickly.
  - Files: `src/renderer/components/chat/MessageInput.tsx`, `src/renderer/components/chat/CommandSuggestionDropdown.tsx`
  - Do:
    1. Ensure Esc closes suggestions without insertion and without triggering send.
    2. Add stable diagnostic codes for acceptance success/no-selection/duplicate suppression (dev-only debug + warning/error boundaries).
    3. Keep slash acceptance behavior isolated from non-slash submit flow.
    4. Run full S03 verification commands and tune assertions if needed.
  - Verify: both new test files pass + `bun run typecheck` passes.
  - Done when: R006 is proven, supported R005/R007 checks are green, and diagnostics are available for failure-path inspection.

## Files Likely Touched

- `src/renderer/components/chat/MessageInput.tsx`
- `src/renderer/components/chat/CommandSuggestionDropdown.tsx`
- `src/renderer/hooks/useCommandSuggestions.ts`
- `src/renderer/components/chat/__tests__/MessageInput.slash-acceptance.test.tsx`
- `src/renderer/components/chat/__tests__/CommandSuggestionDropdown.acceptance.test.tsx`

## Forward Intelligence

(To be populated after execution with fragility notes and rerun guidance.)

### PR DESCRIPTION
**Pull request by gannonh**
## Summary
- add S03 acceptance contract tests for MessageInput and CommandSuggestionDropdown
- implement deterministic Tab/Enter slash suggestion acceptance with trailing-space insertion, focus/caret retention, and Enter-to-send arbitration
- add stable dev diagnostics (`SLASH_ACCEPTED`, `SLASH_ACCEPT_NO_SELECTION`, `SLASH_ACCEPT_SUPPRESSED_DUPLICATE`) and reset duplicate suppression across input edits

## Verification
- `cd apps/desktop && bun run test -- --coverage.enabled=false src/renderer/components/chat/__tests__/MessageInput.slash-acceptance.test.tsx`
- `cd apps/desktop && bun run test -- --coverage.enabled=false src/renderer/components/chat/__tests__/CommandSuggestionDropdown.acceptance.test.tsx`
- `cd apps/desktop && bun run typecheck`
- `cd apps/desktop && bun run build`
- `git push origin desktop/_S03__343` (pre-push hook ran affected lint/typecheck/test pipeline)

Closes #343

## Summary by CodeRabbit

* **New Features**
  * Tab and Enter both accept highlighted slash-command suggestions; pointer selection also accepts suggestions.
  * Tracks recent acceptance to suppress rapid duplicate inserts and emits diagnostics for acceptance events.

* **Bug Fixes**
  * Prevents no-op insertions and unintended focus capture when no suggestion is active.
  * Warns when attempting to accept with no selectable suggestion.

* **Tests**
  * Added acceptance tests covering keyboard, pointer, suppression, warnings, and edge cases.

<h3>Greptile Summary</h3>

This PR finalizes the slash command autocomplete acceptance contract by adding Tab as an acceptance key alongside Enter, introducing a 64 ms duplicate-suppression guard via `lastAcceptedSuggestionRef`, and extracting the acceptance path into a dedicated `acceptSuggestion` callback with dev-only diagnostics. Two new test files provide S03 contract coverage for `MessageInput` and `CommandSuggestionDropdown`.

<h3>Confidence Score: 5/5</h3>

Safe to merge; all remaining findings are P2 style/naming suggestions in dev-only diagnostic code.

The acceptance logic is correct and well-tested. The two P2 findings are limited to misleading debug-log labelling and an undocumented intentional behaviour change around Tab during loading — neither affects runtime correctness or production builds.

MessageInput.tsx — the no-op diagnostic label and Tab-during-loading interception are worth a follow-up clarification comment in code.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| apps/desktop/src/renderer/components/chat/MessageInput.tsx | Core logic refactored: applyCommandSuggestion now returns bool, new acceptSuggestion callback adds dedup ref + diagnostics, Tab added as acceptance key alongside Enter; two P2 notes (misleading diagnostic code and Tab-during-loading behavior). |
| apps/desktop/src/renderer/components/chat/__tests__/MessageInput.slash-acceptance.test.tsx | New acceptance contract tests covering Tab, Enter, Escape, submit-when-closed, no-selection warning, duplicate suppression, and reset-on-value-change; well-structured mock with keepSuggestionsVisibleAfterAccept flag for edge cases. |
| apps/desktop/src/renderer/components/chat/__tests__/CommandSuggestionDropdown.acceptance.test.tsx | New tests verify onSelect invocation, mousedown default-prevention for focus retention, and no-op behavior with empty suggestions; clean setup with ResizeObserver mock. |

</details>

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    K[KeyDown: Enter / Tab / Pointer click] --> SC{showSuggestions?}
    SC -- No --> EN{Enter key?}
    EN -- Yes --> SEND[handleSend]
    EN -- No --> NOOP1[no-op]
    SC -- Yes --> AS[acceptSuggestion key, command?]
    AS --> SEL{selectedSuggestion resolved?}
    SEL -- No --> WARN[emitDiagnostic SLASH_ACCEPT_NO_SELECTION / return]
    SEL -- Yes --> DUP{Within 64ms duplicate?}
    DUP -- Yes --> SUPP[dismiss + emitDiagnostic SLASH_ACCEPT_SUPPRESSED_DUPLICATE / return]
    DUP -- No --> SETREF[set lastAcceptedSuggestionRef]
    SETREF --> APPLY[applyCommandSuggestion]
    APPLY --> NOOP2{nextValue === value?}
    NOOP2 -- Yes --> DISMISS[setIsSuggestionDismissed true / return false]
    NOOP2 -- No --> INSERT[setValue + setSelectedIndex + queueMicrotask focus+caret / return true]
    INSERT --> DIAG1[emitDiagnostic SLASH_ACCEPTED]
    DISMISS --> DIAG2[emitDiagnostic SLASH_ACCEPT_SUPPRESSED_DUPLICATE - misleading label for no-op]
```

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: apps/desktop/src/renderer/components/chat/MessageInput.tsx
Line: 219-229

Comment:
**Misleading diagnostic code for no-op path**

When `applyCommandSuggestion` returns `false` because `nextValue === value` (the input already contains the exact text the suggestion would produce), `SLASH_ACCEPT_SUPPRESSED_DUPLICATE` is emitted. But nothing was duplicated — the insertion was a genuine no-op. This conflates two distinct failure modes under one diagnostic label, making it harder to distinguish a real dedup event from a no-op during debugging. A dedicated code like `SLASH_ACCEPT_NOOP` would keep these cases separable in the console output.

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: apps/desktop/src/renderer/components/chat/MessageInput.tsx
Line: 271-279

Comment:
**Tab intercept blocks focus traversal during command loading**

`showSuggestions` is true whenever `isOpen && (isLoading || suggestions.length > 0)`. This means Tab is now intercepted even while `isLoading && suggestions.length === 0` — the loading-spinner state. A user who types `/` and then immediately presses Tab to move focus elsewhere will find the key silently consumed (with only a `SLASH_ACCEPT_NO_SELECTION` warn emitted). The dropdown is visible during loading so the interception is arguably intentional, but it's a behaviour change from the previous Enter-only path and worth a comment in code to make the intent explicit.

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Reviews (1): Last reviewed commit: ["fix(S03/T03): reset acceptance dedupe ac..."](https://github.com/gannonh/kata/commit/4aa09cf68ad8c66a9d3394e96e4e92bc41a5e630) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=29206663)</sub>

> Greptile also left **2 inline comments** on this PR.

### CODE REVIEW

File: apps/desktop/src/renderer/components/chat/__tests__/MessageInput.slash-acceptance.test.tsx [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+  })
+}
+
+beforeEach(() => {
+  setupSuggestionsMock()
+})
+
+afterEach(() => {
+  cleanup()
+  vi.clearAllMocks()
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/src/renderer/components/chat/__tests__/MessageInput.slash-acceptance.test.tsx]: Good catch — this is now handled. The test file uses `vi.restoreAllMocks()` in `afterEach`, so `console.warn`/`console.debug` spies are restored between tests and no longer leak across cases.

[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/src/renderer/components/chat/__tests__/MessageInput.slash-acceptance.test.tsx]: Updated in current branch: test cleanup now uses `vi.restoreAllMocks()` in `afterEach` to avoid leaked `console` spies across tests.

File: apps/desktop/src/renderer/components/chat/__tests__/MessageInput.slash-acceptance.test.tsx [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+      expect((screen.getByTestId('chat-input') as HTMLTextAreaElement).value).toBe('/kata plan ')
+    })
+
+    expect(debugSpy).toHaveBeenCalledWith(
+      '[SlashAutocomplete]',
+      expect.objectContaining({
+        code: 'SLASH_ACCEPT_SUPPRESSED_DUPLICATE',
+      }),
+    )
+  })
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/src/renderer/components/chat/__tests__/MessageInput.slash-acceptance.test.tsx:297]: Addressed. The duplicate-suppression test now controls `Date.now()` and performs the second acceptance within 20ms of the first acceptance, so it exercises the `SLASH_DUPLICATE_SUPPRESSION_WINDOW_MS` branch directly rather than the no-op branch.

[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/src/renderer/components/chat/__tests__/MessageInput.slash-acceptance.test.tsx:297]: Addressed. The duplicate-suppression test now controls `Date.now()` and performs the second accept within the suppression window, so it deterministically exercises the time-window branch (not the no-op branch).

File: apps/desktop/src/renderer/components/chat/__tests__/MessageInput.slash-acceptance.test.tsx [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+      ([scope, payload]) =>
+        scope === '[SlashAutocomplete]' &&
+        typeof payload === 'object' &&
+        payload !== null &&
+        'code' in payload &&
+        (payload as { code?: string }).code === 'SLASH_ACCEPT_SUPPRESSED_DUPLICATE',
+    )
+
+    expect(suppressedDuplicateEvents).toHaveLength(0)
+  })
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/src/renderer/components/chat/__tests__/MessageInput.slash-acceptance.test.tsx:338]: Addressed. The re-acceptance test now seeds `selectedIndex: 0`, accepts `/kata`, resets input via user edit, and accepts `/kata` again. This keeps the accepted command consistent across both passes and validates that duplicate suppression is reset by real input changes.

File: apps/desktop/src/renderer/components/chat/MessageInput.tsx [hunk 4]
```diff
@@ -118,7 +172,64 @@ export function MessageInput({
...
+        'debug',
+        inserted ? 'SLASH_ACCEPTED' : 'SLASH_ACCEPT_SUPPRESSED_DUPLICATE',
+        {
+          key,
+          prefix: slashPrefix,
+          command: selectedSuggestion.name,
+          selectedIndex,
+          suggestionCount: suggestions.length,
+        },
+      )
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/src/renderer/components/chat/MessageInput.tsx]: Addressed. The no-op path now emits `SLASH_ACCEPT_NO_OP` and `SLASH_ACCEPT_SUPPRESSED_DUPLICATE` is reserved for true suppression-window dedupe events. This keeps diagnostics unambiguous for triage.

File: apps/desktop/src/renderer/components/chat/__tests__/MessageInput.slash-acceptance.test.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/src/renderer/components/chat/__tests__/MessageInput.slash-acceptance.test.tsx:338]: Addressed. The re-acceptance test now keeps the same suggestion selected across both accepts and asserts no `SLASH_ACCEPT_SUPPRESSED_DUPLICATE` emission after input changes.

File: apps/desktop/src/renderer/components/chat/MessageInput.tsx [hunk 5]
```diff
@@ -157,14 +268,13 @@ export function MessageInput({
...
               event.preventDefault()
               event.stopPropagation()
-
-              const selectedSuggestion = suggestions[selectedIndex]
-              if (selectedSuggestion) {
-                applyCommandSuggestion(selectedSuggestion)
-              }
+              acceptSuggestion(shouldAcceptWithTab ? 'Tab' : 'Enter')
               return
             }
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/src/renderer/components/chat/MessageInput.tsx:328]: Handled in the current key arbitration: Tab acceptance is now gated by `hasSelectableSuggestion`, so Tab is not consumed during loading/empty suggestion states and normal focus traversal is preserved.

File: apps/desktop/src/renderer/components/chat/MessageInput.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/src/renderer/components/chat/MessageInput.tsx]: Addressed in current code path: no-op acceptance now emits `SLASH_ACCEPT_NO_OP` instead of `SLASH_ACCEPT_SUPPRESSED_DUPLICATE`, keeping diagnostics distinct.

File: apps/desktop/src/renderer/components/chat/MessageInput.tsx [hunk 6]
```diff
@@ -157,14 +268,13 @@ export function MessageInput({
...
+
+            if (showSuggestions && (shouldAcceptWithEnter || shouldAcceptWithTab)) {
               event.preventDefault()
               event.stopPropagation()
-
-              const selectedSuggestion = suggestions[selectedIndex]
-              if (selectedSuggestion) {
-                applyCommandSuggestion(selectedSuggestion)
-              }
+              acceptSuggestion(shouldAcceptWithTab ? 'Tab' : 'Enter')
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/src/renderer/components/chat/MessageInput.tsx]: Addressed by the updated Enter/Tab guards. Acceptance now only intercepts when a selectable suggestion exists; otherwise Enter falls back to normal send behavior and Tab falls back to native focus navigation.

File: apps/desktop/src/renderer/components/chat/MessageInput.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/src/renderer/components/chat/MessageInput.tsx:328]: Fixed. Tab acceptance is now gated by `hasSelectableSuggestion`, so Tab is no longer intercepted during loading/empty states and normal focus traversal is preserved.

File: apps/desktop/src/renderer/components/chat/MessageInput.tsx [hunk 7]
```diff
@@ -157,14 +298,22 @@ export function MessageInput({
...
+            const shouldAcceptWithTab = event.key === 'Tab' && !event.shiftKey
+            const hasSelectableSuggestion =
+              suggestions.length > 0 && selectedIndex >= 0 && selectedIndex < suggestions.length
+
+            if (showSuggestions && shouldAcceptWithEnter) {
               event.preventDefault()
               event.stopPropagation()
+              acceptSuggestion('Enter')
+              return
+            }
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/src/renderer/components/chat/MessageInput.tsx:311]: Fixed. Enter acceptance now requires `hasSelectableSuggestion`; when suggestions are open but unselectable (loading/empty/out-of-range), Enter falls through to normal submit behavior. Added coverage in `MessageInput.slash-acceptance.test.tsx` (`Enter falls back to submit when suggestions are visible but no selection is available`).

File: apps/desktop/src/renderer/components/chat/MessageInput.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/src/renderer/components/chat/MessageInput.tsx]: Addressed. Enter/Tab now fall through when no selectable suggestion is present; acceptance handling is only invoked when a suggestion is actually selectable.

File: apps/desktop/src/renderer/components/chat/MessageInput.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/src/renderer/components/chat/MessageInput.tsx:311]: Implemented. Enter arbitration now uses `hasSelectableSuggestion` before intercepting, so when the dropdown is visible but empty/loading Enter falls through to normal submit behavior. We still emit `SLASH_ACCEPT_NO_SELECTION` as diagnostics without swallowing the send path.
